File: 409/Solution.py
def longestPalindrome(s: str) -> int:
    # 用字典统计每个字符的出现次数
    char_count = {}
    for char in s:
        char_count[char] = char_count.get(char, 0) + 1

    # 计算最长回文串的长度
    palindrome_length = 0
    odd_found = False
    for count in char_count.values():
        # 如果是偶数，全部加到回文长度
        if count % 2 == 0:
            palindrome_length += count
        else:
            # 如果是奇数，取偶数部分
            palindrome_length += count - 1
            odd_found = True

    # 如果有奇数个字符，可以放一个到中心
    if odd_found:
        palindrome_length += 1

    return palindrome_length

# 测试
s = "abccccdd"
print(longestPalindrome(s))  # 输出 7

File: 409/Solution.cpp
/*
给定一个包含大写字母和小写字母的字符串 s ，返回 通过这些字母构造成的 最长的 回文串 的长度。

在构造过程中，请注意 区分大小写 。比如 "Aa" 不能当做一个回文字符串。
*/

#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int longestPalindrome(string s) {
        int map[52] = {0};
        for (char c : s) {
            if (c >= 'a' && c <= 'z') {
                map[c - 'a' + 26]++;  // 小写字母映射到26-51
            } else if (c >= 'A' && c <= 'Z') {
                map[c - 'A']++;       // 大写字母映射到0-25
            }
            // 忽略非字母字符（根据题目要求可以添加）
        }
        int length = 0;
        bool hasOdd = false;
        
        for (int count : map) {
            if (count % 2 == 0) {
                length += count;
            } else {
                length += count - 1;
                hasOdd = true;
            }
        }
        
        return length + (hasOdd ? 1 : 0);
    }
};
